### ISSUE TITLE
async client reconnect

### ISSUE THREAD

**Issue by srbpavel** [eclipse/paho.mqtt.rust#146]
Dear, please with blocking sync Client i'am able to ```client.reconnect().is_ok()``` when testing mosquitto broker restart or what ever.

But with async AsyncClient and token i did not make it recover:

i have tested it also with your mqtt_chat example as following:

```
cli.set_connection_lost_callback(|client| {
    let mut reconnect_counter = 0;

    loop {

        reconnect_counter += 1;
        println!("reconnect counter: {reconnect_counter}");

         let server_response = client
            .reconnect()
               .try_wait();

           println!("ServerResponse: {server_response:?}");
```
this is what i receive: ```Some(Err(PahoDescr(-18, "Unknown Error")))```

my settings is the same as in your example. tcp protocol + mqttv 5 .... and only added  user/pass in ConnectOptionsBuilder

did i understand the reconnect in async client correctly?

thank you pavel

**Comment by Tockra** [comment]
Did you find a solution?

**Comment by srbpavel** [comment]
> Did you find a solution?

honestly i do not remember, probably i have choose a different way. this was my first try with mqtt_paho
https://github.com/srbpavel/rust/blob/main/mqtt_client/src/lib.rs

**Comment by fpagliughi** [comment]
Sorry. It looks like a bug in the upstream Paho C library is causing havoc on reconnect, which send to result in bad pointer being passed around. It manifests in a number of ways, including segfaults, double free errors, and this, apparently.

The next release will get it fixed.

**Comment by mushonnip** [comment]
> Sorry. It looks like a bug in the upstream Paho C library is causing havoc on reconnect, which send to result in bad pointer being passed around. It manifests in a number of ways, including segfaults, double free errors, and this, apparently.
> 
> The next release will get it fixed.

Has this been fixed now? I had the same experience. I hope the next version will be released soon.

**Comment by fpagliughi** [comment]
It looks like this is fixed with the upcoming v0.12.3 release. At least I can't recreate it in testing.
If you still see it with that version, please feel free to reopen.
